iroh-bytes/src/util/sparse_mem_file.rs

```diff
@@ -0,0 +1,122 @@
+use std::io;
+
+use bao_tree::io::sync::{ReadAt, Size, WriteAt};
+use derive_more::Deref;
+use range_collections::{range_set::RangeSetRange, RangeSet2};
+
+/// A file that is sparse in memory
+///
+/// It is not actually using sparse storage to make reading faster, so it will
+/// not conserve memory. It is just a way to remember the gaps so we can
+/// write it to a file in a sparse way later.
+#[derive(derive_more::Debug)]
+pub struct SparseMemFile {
+    /// The data, with gaps filled with zeros
+    #[debug("{} bytes", data.len())]
+    data: Vec<u8>,
```

Comment:
maybe use `Box<[u8]>`, or can you resize this?

Reply:
No, it can grow. The place where it grows is a bit hidden. In `write_at` I use the WriteAt instance of Vec<u8> that will extend the vec if you write behind the end.